k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapConnection.java

```diff
@@ -356,6 +371,75 @@ private void authenticate() throws MessagingException, IOException {
         }
     }
 
+    private void authXoauth2withSASLIR() throws IOException, MessagingException {
+        retryXoauth2WithNewToken = true;
+        try {
+            attemptXOAuth2();
+        } catch (NegativeImapResponseException e) {
+            //We couldn't login with the token so invalidate it
+            oauthTokenProvider.invalidateToken(settings.getUsername());
+
+            if (!retryXoauth2WithNewToken) {
+                handlePermanentXoauth2Failure(e);
+            } else {
+                handleTemporaryXoauth2Failure(e);
+            }
+        }
+    }
+
+    private void handlePermanentXoauth2Failure(NegativeImapResponseException e) throws AuthenticationFailedException {
+        Log.v(LOG_TAG, "Permanent failure during XOAUTH2", e);
+        throw new AuthenticationFailedException(e.getMessage(), e);
+    }
+
+    private void handleTemporaryXoauth2Failure(NegativeImapResponseException e) throws IOException, MessagingException {
+        //We got a response indicating a retry might suceed after token refresh
+        //We could avoid this if we had a reasonable chance of knowing
+        //if a token was invalid before use (e.g. due to expiry). But we don't
+        //This is the intended behaviour per AccountManager
+
+        Log.v(LOG_TAG, "Temporary failure - retrying with new token", e);
+        try {
+            attemptXOAuth2();
+        } catch (NegativeImapResponseException e2) {
+            //Okay, we failed on a new token.
+            //Invalidate the token anyway but assume it's permanent.
+            Log.v(LOG_TAG, "Authentication exception for new token, permanent error assumed", e);
+            oauthTokenProvider.invalidateToken(settings.getUsername());
+            handlePermanentXoauth2Failure(e2);
+        }
+    }
+
+    private void attemptXOAuth2() throws MessagingException, IOException {
+        String token = oauthTokenProvider.getToken(settings.getUsername(),
+                OAuth2TokenProvider.OAUTH2_TIMEOUT);
+        String tag = sendSaslIrCommand(Commands.AUTHENTICATE_XOAUTH2,
+                Authentication.computeXoauth(settings.getUsername(), token), true);
+
+        extractCapabilities(
+            responseParser.readStatusResponse(tag, Commands.AUTHENTICATE_XOAUTH2, getLogId(),
+                    new UntaggedHandler() {
+                @Override
+                public void handleAsyncUntaggedResponse(ImapResponse response) throws IOException {
+                    handleXOAuthUntaggedResponse(response);
+
+                }
+            })
+        );
+    }
+
+    private void handleXOAuthUntaggedResponse(ImapResponse response) throws IOException {
+        if (response.isString(0)) {
+            retryXoauth2WithNewToken = XOAuth2ChallengeParser.shouldRetry(
+                    response.getString(0), settings.getHost());
+        }
+
+        if(response.isContinuationRequested()) {
```

Comment:
If it was me I'd have a commit that fixed all the errors and made the build fail but nobody's a fan of that :(

I'll fix it.